src/lib/mina_net2/bitswap_block.ml

```diff
@@ -0,0 +1,237 @@
+open Core_kernel
+
+let or_error_list_bind ls ~f =
+  let open Or_error.Let_syntax in
+  let rec loop = function
+    | [] ->
+        return []
+    | h :: t ->
+        let%bind r = f h in
+        let%map t' = loop t in
+        r :: t'
+  in
+  loop ls >>| List.concat
+
+type link = Blake2.t
+
+let link_size = Blake2.digest_size_in_bytes
+
+let absolute_max_links_per_block = Stdint.Uint16.(to_int max_int)
+
+type schema =
+  { total_blocks : int
+  ; full_link_blocks : int
+  ; non_max_link_block_count : int
+  ; last_block_data_size : int
+  ; max_block_size : int
+  ; max_links_per_block : int
+  }
+[@@deriving compare, eq, sexp]
+
+let required_bitswap_block_count ~max_block_size data_length =
+  if data_length <= max_block_size - 2 then 1
+  else
+    let n1 = data_length - link_size in
+    let n2 = max_block_size - link_size - 2 in
+    (n1 / n2) + if n1 mod n2 > 0 then 1 else 0
+
+let max_links_per_block ~max_block_size =
+  let links_per_block = (max_block_size - 2) / link_size in
+  min links_per_block absolute_max_links_per_block
+
+let create_schema ~max_block_size data_length =
+  let total_blocks = required_bitswap_block_count ~max_block_size data_length in
+  let last_block_data_size =
+    data_length - ((max_block_size - link_size - 2) * (total_blocks - 1))
+  in
+  let max_links_per_block = max_links_per_block ~max_block_size in
+  let non_max_link_block_count = (total_blocks - 1) mod max_links_per_block in
```

Comment:
It would be more intuitive (for me, at least) if this was calculated after `full_link_blocks` as
```ocaml
let non_max_link_block_count =
  total_blocks - full_link_blocks * max_links_per_block
in
```